### PR TITLE
Manually triggered GitHub Action for static checks using clang-tidy

### PR DESCRIPTION
**Pull request by williamfgc**
## Proposed changes

Adds a `static` manual workflow on GitHub Actions, triggered from the Actions tab. 
Get results from compiling with clang-tidy (takes 2 hours 15 min) on-demand on the develop branch only.
Update docker image to add clang-tidy dependency.
Related to #3941 as we address clang-tidy recommendations.
Needs merging so an admin can trigger the workflow.

## What type(s) of changes does this code introduce?

- Testing changes (e.g. new unit/integration/performance tests) static checks

### Does this introduce a breaking change?

- Yes

## What systems has this change been tested on?
Runs on GitHub Actions on my fork, must run on QMCPACK/qmcpack

## Checklist

- Yes. This PR is up to date with current the current state of 'develop'
- No. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by williamfgc** [comment]
For reference this is the resulting workflow logs on my [fork](https://github.com/williamfgc/qmcpack/actions/runs/2339592357).

**Comment by correaa** [comment]
The PR looks good to me. 
Depending on the number of compilation units, `"clang-tidy;-checks=*"` might produce very large logs to the point that github could truncate them.

There will be a flood of warnings at the beginning so we might want to define a protocol on how to deal with them, for example if it is decided that nothing will be done about a certain warning (for example, disable the warning globally vs. disabling check in that line `  // NOLINT` vs disabling specific check `  // NOLINT(specific-warning)` vs adding a useful comment ` // NOLINT(specific-warning) comment: implicit conversion is necessary here because bla...`.

I think we can safely disable this three warnings to start with because they don't make sense.

```
  -llvmlibc-callee-namespace,
  -llvmlibc-implementation-in-namespace,
  -llvmlibc-restrict-system-libc-headers,
```

Finally, as exceptions mount, it is better to have `.clang-tidy` configuration file that can sit in the root directory. For example: https://gitlab.com/correaa/boost-multi/-/blob/master/.clang-tidy

**Review by correaa — APPROVED** [review]
Look good to me. See comments in the PR thread about possible future issues.

**Comment by ye-luo** [comment]
Need some documentation. I mean if I see a CI failure, how to reproduce clang-tidy manually.

**Comment by correaa** [comment]
> Need some documentation. I mean if I see a CI failure, how to reproduce clang-tidy manually.



> Need some documentation. I mean if I see a CI failure, how to reproduce clang-tidy manually.

yes, even for reproducibility sake, as soon as the command line becomes more complicated that `-DCMAKE_CXX_CLANG_TIDY="clang-tidy;-checks=*"`, we will start needing a configuration file `.clang-tidy` that is part of the repository.

**Comment by markdewing** [comment]
There already is a .clang-tidy file in the src/ directory
https://github.com/QMCPACK/qmcpack/blob/develop/src/.clang-tidy

**Comment by correaa** [comment]
> There already is a .clang-tidy file in the src/ directory https://github.com/QMCPACK/qmcpack/blob/develop/src/.clang-tidy

ok, in that case the command in the CI can be simply `-DCMAKE_CXX_CLANG_TIDY="clang-tidy"`.

**Review by prckent — CHANGES_REQUESTED** [review]
We should be using a current version of Clang right from the beginning. e.g. 13, since 14 is a bit too recent to be available everywhere. LLVM10 will be out of our 2 year support window in 4 months.

Given that we have so many warnings at the moment, have you found a convenient way to review them in the GitHub interface? My experience with the current CI logs is not good for any lengthy file.

**Comment by williamfgc** [comment]
Changes:
- Added docs @ye-luo 
- Added Ubuntu22 image with Clang14 (default) for a recent clang-tidy @prckent 
- Added `.clang-tidy` with `readability-*` and `bugprone-*` as seen in #3941. @correaa

File: docs/github_actions.rst [hunk 1]
```diff
@@ -945,3 +945,30 @@ Generate code coverage reports once all tests have reported.
...
+
+To run the workflow:
+- Go to the Actions tab
+- Click on the `static` workflow on the left
+- Click on `Run workflow` on the right
+- Use workflow from `Branch:develop` and click on the `Run workflow` button
+
+**Note:** the current `.clang-tidy` configuration file is compatible with clang v14 and runs on the `williamfgc/qmcpack-ci:ubuntu22-openmpi` docker image. To run locally on a Linux system use: `docker run -it user williamfgc/qmcpack-ci:ubuntu22-openmpi /bin/bash` or refer to the :ref:`running_docker`` section.
+
+To build locally enaling `clang-tidy`` static checks defined in `qmcpack/.clang-tidy` use the CMake `-DCMAKE_CXX_CLANG_TIDY` option as follows:
```
Comment (ye-luo) [inline docs/github_actions.rst:964]: `src/.clang-tidy` instead of `qmcpack/.clang-tidy`

Comment (williamfgc) [reply docs/github_actions.rst:964]: @ye-luo good point. @PDoakORNL @quantumsteve I'm unifying the `.clang-tidy` configuration. Are you OK removing the one under `/src`? The one at the top level would have similar checks.

Comment (ye-luo) [reply docs/github_actions.rst:964]: We should have only one configure file. I don't get why it should be under the root directory. If I understand correctly, clang-tidy handles C++ source code and all C++ code managed by QMCPACK is under src. Codes under external should not be subject to clang-tidy checks by QMCPACK.

File: docs/github_actions.rst [hunk 2]
```diff
@@ -945,3 +945,30 @@ Generate code coverage reports once all tests have reported.
...
+
+To run the workflow:
+- Go to the Actions tab
+- Click on the `static` workflow on the left
+- Click on `Run workflow` on the right
+- Use workflow from `Branch:develop` and click on the `Run workflow` button
+
+**Note:** the current `.clang-tidy` configuration file is compatible with clang v14 and runs on the `williamfgc/qmcpack-ci:ubuntu22-openmpi` docker image. To run locally on a Linux system use: `docker run -it user williamfgc/qmcpack-ci:ubuntu22-openmpi /bin/bash` or refer to the :ref:`running_docker`` section.
+
+To build locally enaling `clang-tidy`` static checks defined in `qmcpack/src/clang-tidy` use the CMake `-DCMAKE_CXX_CLANG_TIDY` option as follows:
```
Comment (ye-luo) [inline docs/github_actions.rst:964]: `qmcpack/src/.clang-tidy` missing a dot.
"enaling" typo?

**Comment by PDoakORNL** [comment]
Our .clang-tidy should be in src it should not be a reachable parent for the external_codes.  Additionally be aware that subdirectories can have there own .clang-tidy which may optionally inherit from its parent or not.  For instance in src/type_traits if am experimenting with the following 
```yaml
InheritParentConfig: true
CheckOptions: [ { key: readability-identifier-naming.StructIgnoredRegexp, value: '.*(_t|_impl)$'}
                ]
```
since its rather helpful to reading template metaprogramming to have types with _t and _impl and no have a ton of //NOLINT comments.

Some of the lost cause directories could also be squelched from the clang-tidy reporting by adding `InheritParentConfig: false` .clang-tidy in their directory.

**Comment by correaa** [comment]
> since its rather helpful to reading template metaprogramming to have types with _t and _impl and no have a ton of //NOLINT comments.

I use that idiom in my naming and yet I didn't have to put anything special to the clangtidy.


> Some of the lost cause directories could also be squelched from the clang-tidy reporting by adding `InheritParentConfig: false` in .clang-tidy local to those directories.  

I didn't know about this option, seems useful (but not for external code) but for directories, (like tests).

**Comment by PDoakORNL** [comment]
@correaa in our case we've defined type naming rules that disallow that for types.  They're specified as JustMixedCase.

**Comment by williamfgc** [comment]
Thank you all for your input. I think I addressed the review comments. Let me know if there anything pending. We can all refer to the manually generated report after merge to start addressing issues, in particular those of `bugprone-*` type.

File: docs/github_actions.rst [hunk 3]
```diff
@@ -945,3 +945,30 @@ Generate code coverage reports once all tests have reported.
 Upload Coverage
 ---------------
 Upload the generated code coverage to `CodeCov <https://codecov.io/gh/QMCPACK/qmcpack/tree/develop/src>`_ where the badges on our repo will then be updated. Only done by jobs with name `*-coverage`.
+
+
+====== ======== ======== 
+Static Analysis Workflow
+====== ======== ========
```
Comment (ye-luo) [inline docs/github_actions.rst:952]: I don't see breaking lines based on spaces.

File: docs/github_actions.rst [hunk 4]
```diff
@@ -945,3 +945,30 @@ Generate code coverage reports once all tests have reported.
...
+
+To run the workflow:
+- Go to the Actions tab
+- Click on the `static` workflow on the left
+- Click on `Run workflow` on the right
+- Use workflow from `Branch:develop` and click on the `Run workflow` button
+
+**Note:** the current `.clang-tidy` configuration file is compatible with clang v14 and runs on the `williamfgc/qmcpack-ci:ubuntu22-openmpi` docker image. To run locally on a Linux system use: `docker run -it user williamfgc/qmcpack-ci:ubuntu22-openmpi /bin/bash` or refer to the :ref:`running_docker`` section.
+
+To build locally enaling `clang-tidy`` static checks defined in `qmcpack/src/.clang-tidy` use the CMake `-DCMAKE_CXX_CLANG_TIDY` option as follows:
```
Comment (ye-luo) [inline docs/github_actions.rst:964]: `enaling` is still there.

File: docs/github_actions.rst [hunk 5]
```diff
@@ -945,3 +945,30 @@ Generate code coverage reports once all tests have reported.
 Upload Coverage
 ---------------
 Upload the generated code coverage to `CodeCov <https://codecov.io/gh/QMCPACK/qmcpack/tree/develop/src>`_ where the badges on our repo will then be updated. Only done by jobs with name `*-coverage`.
+
+
+====== ======== ======== 
+Static Analysis Workflow
+====== ======== ========
+
+A manually triggered workflow on the GitHub Actions tab can generate the required checks using the [clang-tidy](https://clang.llvm.org/extra/clang-tidy/) static analyzer. The current approach is to set checks in the `qmcpack/.clang-tidy` configuration file and run using `clang-tidy` v14 on GitHub Actions runners. The workflow is not part of CI, and it's currently used for reporting potential warnings on the GitHub Actions logs as they are addressed on the `develop` branch as part of refactoring efforts for code quality.
```
Comment (ye-luo) [inline docs/github_actions.rst:954]: `qmcpack/.clang-tidy` still

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — DISMISSED** [review]

**Review by prckent — APPROVED** [review]

**Comment by prckent** [comment]
Approved assuming that I didn't break the rst.

**Comment by prckent** [comment]
Test this please

**Review by prckent — APPROVED** [review]

**Comment by correaa** [comment]
ok, i see. i didn’t know about JustMixedCase. what is the difference with
CamelCase?

**Pull request merged**